Build Tools: Copy block-library PHP metadata files with PHP source files

chipsnyder opened the pull request:
Fixes: https://core.trac.wordpress.org/ticket/49196
Related to: https://github.com/WordPress/gutenberg/pull/19534

Looks to see if there is a `block.json` file in the same directory as the block's defined in `phpfiles` then copies those to the same directory.

gziolo commented on tools/webpack/packages.js:
```diff
@@ -152,6 +152,16 @@ module.exports = function( env = { environment: 'production', watch: false, buil
+	const phpMetadataCopies = Object.keys( phpFiles ).map( function( filename ) {
+		return {
+			filename: filename,
+			dir:dirname( filename ),
```
Nit, space is missing:
```suggestion
			dir: dirname( filename ),
```

gziolo commented on tools/webpack/packages.js:
```diff
@@ -152,6 +152,16 @@ module.exports = function( env = { environment: 'production', watch: false, buil
+			filename: filename,
+			dir:dirname( filename ),
+		  }
+	} ).map( ({ filename, dir }) => ( {
```
Nit, spaces are missing:
```suggestion
	} ).map( ( { filename, dir } ) => ( {
```

gziolo commented on tools/webpack/packages.js:
```diff
@@ -152,6 +152,16 @@ module.exports = function( env = { environment: 'production', watch: false, buil
 		to: join( baseDir, `src/${ phpFiles[ filename ] }` ),
 	} ) );
 
+	const phpMetadataCopies = Object.keys( phpFiles ).map( function( filename ) {
```
It's a block metadata.
```suggestion
	const blockMetadataCopies = Object.keys( phpFiles ).map( function( filename ) {
```

gziolo commented on tools/webpack/packages.js:
```diff
@@ -152,6 +152,16 @@ module.exports = function( env = { environment: 'production', watch: false, buil
+		  }
+	} ).map( ({ filename, dir }) => ( {
+		from: join( baseDir, `node_modules/@wordpress/${ dir }/block.json` ),
+		to: join( baseDir, `src/${ dirname(phpFiles[ filename ]) }/${ basename(dir) }.json` ),
```
Nit, spaces are missing:
```suggestion
		to: join( baseDir, `src/${ dirname( phpFiles[ filename ] ) }/${ basename( dir ) }.json` ),
```

gziolo commented:
Thank you for opening this PR. It should be filed in WordPress trac:
https://core.trac.wordpress.org

This is only a mirror of svn repository and it's useful for testing against Continues Integration.

chipsnyder commented on tools/webpack/packages.js:
```diff
@@ -231,6 +241,7 @@ module.exports = function( env = { environment: 'production', watch: false, buil
 					...vendorCopies,
 					...cssCopies,
 					...phpCopies,
+					...phpMetadataCopies,
```
```suggestion
					...blockMetadataCopies,
```

chipsnyder commented:
Thanks for taking the time to look at it @gziolo. I updated the PR based on the suggestions and updated the diff in the ticket as well. I had failed to read the contributing guide close enough which is why I opened this PR first then added it to the ticket. I can close it after you get a chance to look at the changes :)

gziolo commented:
Can you have a look at https://core.trac.wordpress.org/attachment/ticket/49196/49196-block-folders.diff which I prepared to make it work without errors when running `npm run build:dev` in WordPress core?

If you are fine with the changes proposed, it would be great to validate them on Travis in this branch.

gziolo commented on Gruntfile.js:
```diff
@@ -99,6 +99,9 @@ module.exports = function(grunt) {
 			files: buildFiles.map( function( file ) {
 				return BUILD_DIR + file;
 			}),
+			blocks: [
```
I'm wondering whether this clean task is necessary. We didn't have it before and it was working properly.

chipsnyder replied:
I was just looking into this actually. I just ran a test with that clean and the blocks folder was deleted and I ran again and the folder was preserved. There seems to be an issue here though. I'm seeing that the `{blockname}.php` is being created as well as the `{blockname}/index.php`

chipsnyder replied:
Removed that file. Seems like those extra files came over with `copy:files` I update this PR with some more changes. We can update the task with the new diff if it passes

gziolo replied:
Cool, Travis is happy now. So it looks like unit tests depend on the existence of those files. It's unclear to me why they get removed in the first place. Regardless, it follows what was here before so I can look into it separately.

chipsnyder replied:
> It's unclear to me why they get removed in the first place

I think the `clean:blocks` step was executing after the copy steps. So it would copy then delete them.

chipsnyder commented:
Reopening with new changes
